library_management/src/main/java/com/group2/library_management/repository/specification/BorrowingReceiptSpecification.java

```diff
@@ -0,0 +1,33 @@
+package com.group2.library_management.repository.specification;
+
+import com.group2.library_management.entity.BorrowingReceipt;
+import com.group2.library_management.entity.User;
+import com.group2.library_management.entity.enums.BorrowingStatus;
+import jakarta.persistence.criteria.Predicate;
+import org.springframework.data.jpa.domain.Specification;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BorrowingReceiptSpecification {
+    public static Specification<BorrowingReceipt> filterByCriteria(
+            User user, List<BorrowingStatus> statuses, LocalDate fromDate, LocalDate toDate) {
+        return (root, query, criteriaBuilder) -> {
+            List<Predicate> predicates = new ArrayList<>();
+            
+            predicates.add(criteriaBuilder.equal(root.get("user"), user));
+            
+            if (statuses != null && !statuses.isEmpty()) {
+                predicates.add(root.get("status").in(statuses));
+            }
+            if (fromDate != null) {
+                predicates.add(criteriaBuilder.greaterThanOrEqualTo(root.get("createdAt"), fromDate.atStartOfDay()));
+            }
+            if (toDate != null) {
+                predicates.add(criteriaBuilder.lessThanOrEqualTo(root.get("createdAt"), toDate.atTime(23, 59, 59)));
```

Comment:
fromDate dùng atStartOfDay, còn toDate lại dùng atTime(23,59,59). Nếu hệ thống dùng kiểu timestamp có thể gây sai lệch về múi giờ hoặc mất chính xác ở mili giây.
=> Giải pháp: Nên dùng toDate.plusDays(1).atStartOfDay() để lấy hết ngày toDate, hoặc chuẩn hóa kiểu dữ liệu lưu trữ.

---
Cách 1: Truyền Map để lọc động, dễ mở rộng, phù hợp với form lọc nhiều trường.
Cách 2: Dùng Builder Pattern để code gọn, an toàn kiểu dữ liệu, dễ bảo trì.

chọn đi nhé bấy bì